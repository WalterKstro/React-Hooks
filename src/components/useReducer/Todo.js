import { useEffect, useReducer } from "react"
import todoReducer from "./todoReducer"
import TodosList from '../Todos/TodosList'
import TodoAdd from "../Todos/TodoAdd"

import './style.css'


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) !== null ? JSON.parse(localStorage.getItem('todos')) : []
}
const Todo = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init)


    useEffect(() => {
        const data = JSON.stringify(todos)
        window.localStorage.setItem('todos', data)
    }, [todos])


    const handleAdd = todo => {
        dispatch({
            type: 'add',
            payload : todo
        })
    }
    const handleDelete = id => {
        dispatch({
            type: 'delete',
            payload: id
        })
    }

    const handleToogle = id => {
        dispatch({
            type: 'toogle',
            payload: id
        })
    }

    return (
        <>
         <h2 className="text-center my-5">Todos App ({todos.length})</h2>
         
         <TodoAdd handleAdd={handleAdd}/>
         
         <TodosList
            todos={todos}
            handleToogle={handleToogle}
            handleDelete={handleDelete}
         />
        </>
    )
}

export default Todo
