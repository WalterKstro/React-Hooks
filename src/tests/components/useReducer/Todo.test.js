import { mount } from "enzyme"
import { act } from "react-dom/test-utils"
import Todo from "../../../components/useReducer/Todo"
import state from "./dataTodos"

describe('Testing component <Todo/>', () => {
    let component
    beforeEach(() => {
        component = mount(<Todo/>)
    })

    Storage.prototype.setItem = jest.fn()

    test('should show correctly the component <Todo/>', () => {
        expect(component).toMatchSnapshot()
    })

    test('should add new TODO', () => {
        const handleAdd = component.find('TodoAdd').prop('handleAdd')
        act(() => {
            handleAdd(state[0])
            handleAdd(state[1])
        })

        const todos = component.find('h2').text()
        
        expect(todos).toBe(`Todos App (${state.length})`)
        expect(localStorage.setItem).toBeCalledTimes(state.length)

    })

    test('should delete one TODO', () => {
        const handleAdd = component.find('TodoAdd').prop('handleAdd')
        const handleDelete = component.find('TodosList').prop('handleDelete')
        act(() => {
            handleAdd(state[0])
            handleAdd(state[1])
            handleDelete(state[0].id)
        })

        const todos = component.find('h2').text()
        
        expect(todos).toBe(`Todos App (1)`)
    })
    
    
    
})
