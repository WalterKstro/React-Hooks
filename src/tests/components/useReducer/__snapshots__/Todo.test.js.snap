// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component <Todo/> should show correctly the component <Todo/> 1`] = `
Array [
  <h2
    className="text-center my-5"
  >
    Todos App (
    0
    )
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control mb-3"
      name="task"
      onChange={[Function]}
      placeholder="Quiero aprender ..."
      type="text"
      value=""
    />
  </form>,
  <ol
    className="list-group list-group-numbered"
  />,
]
`;
